resolve_stream.py

```diff
@@ -93,6 +101,112 @@ def print_parameter(param, stream='', rd='', indent_level=0):
         parameter=param, stream=stream, rd=rd, indent=indent * indent_level)
 
 
+def print_qparameter(qparam, indent_level=0):
+    print_parameter(qparam.parameter, qparam.stream, qparam.refdes, indent_level=indent_level)
+
+
+class InvalidParameter(NameError):
+    """
+    Raised when parameter is invalid.
+
+    This occurs when attempting to resolve a parameter for a stream in which it does not belong.
+    """
+
+
+class UnresolvedParameter(UserWarning):
+    """
+    Indicates inability to resolve a parameter.
+
+    This occurs when the preload definition for a parameter is incorrect or when an expected stream
+    definition is missing.
+
+    Note - This does NOT necessary indicate an error. It is possible that a parameter may request
+    another parameter as an optional input.
+    """
+
+
+def resolve_parameter(refdes, method, stream, param, stream_map, m2m, depth=1):
+    """
+    Determine the primary set of parameters required to calculate a derived parameter.
+
+    This only fetches the initial set of parameters. If the initial set of parameters includes
+    functions, they will also need to be resolved with subsequent calls to this function.
+
+    :param refdes:  Fully qualified reference designator for instrument collecting the data.
+    :param method:  Manner in which data is collected (e.g. 'telemetered').
+    :param stream:  Stream name in which the parameter resides (e.g. ctdpf_sbe43_sample).
+    :param param:  Parameter to resolve (e.g. Parameter(seawater_temperature)).
+    :param stream_map:
+    :param m2m:  OOI asset management machine to machine credential object.
+    :return: The set of qualified parameters or None if the parameter is not a derived data product.
+    :throws: ParameterException if input parameter is invalid or if unable to resolve all required parameters
+    """
+    required = []
+    s = Stream.query.filter(Stream.name == stream).first()
+    depth_variance = 17 if 'METBK' in refdes else 6
+    needs = s.needs
+    if needs:
+        this = {refdes}
+        same = same_node(refdes, m2m)
+        collocated = get_collocated(refdes).difference(same).difference(this)
+        near = get_within(refdes, depth_variance).difference(collocated).difference(same).difference(this)
+    else:
+        same = set()
+        collocated = set()
+        near = set()
+
+    mine = make_parameter_map([refdes], method, stream_map)
+    same = make_parameter_map(same, method, stream_map)
+    collocated = make_parameter_map(collocated, method, stream_map)
+    near = make_parameter_map(near, method, stream_map)
```

Comment:
The sets mine, same, collocated, and near are not used unless param.is_function is true, but they are computed before this check. This seems wasteful. I'm not sure if the time lost amounts to anything significant.

Reply:
Those maps have to be set before looping over the stream parameters (each of which will be checked to see if it's a function before resolving). It would unnecessarily complicate the code to only populate them on an as needed basis, but I'm open if you have a clever solution.